pkg/streamingpromql/operators/aggregations/topkbottomk/instant_query.go

```diff
@@ -0,0 +1,338 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package topkbottomk
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"math"
+	"unsafe"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/aggregations"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/pool"
+)
+
+// InstantQuery implements topk() and bottomk() for range queries.
+type InstantQuery struct {
+	Inner                    types.InstantVectorOperator
+	Param                    types.ScalarOperator
+	TimeRange                types.QueryTimeRange
+	Grouping                 []string // If this is a 'without' aggregation, New will ensure that this slice contains __name__.
+	Without                  bool
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	IsTopK                   bool // If false, this is operator is for bottomk().
+
+	expressionPosition posrange.PositionRange
+	k                  int64 // Maximum number of values to return for each group.
+
+	values          []float64
+	nextSeriesIndex int
+
+	annotations                            *annotations.Annotations
+	haveEmittedIgnoredHistogramsAnnotation bool
+
+	// Reuse the same heap instance to allow us to avoid allocating a new one every time.
+	heap *instantQueryHeap
+}
+
+var _ types.InstantVectorOperator = &InstantQuery{}
+
+func (t *InstantQuery) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	if err := t.getK(ctx); err != nil {
+		return nil, err
+	}
+
+	if t.k == 0 {
+		// We can't return any series, so stop now.
+		return nil, nil
+	}
+
+	innerSeries, err := t.Inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(innerSeries)
+
+	groupLabelsBytesFunc := aggregations.GroupLabelsBytesFunc(t.Grouping, t.Without)
+	groups := map[string]*instantQueryGroup{}
+	seriesToGroups := make([]*instantQueryGroup, 0, len(innerSeries))
+
+	// Go through each series and find / create its group, and keep track of how many series contribute to each group.
+	// We do this separately to the loop below so that we know the number of series in each group when we allocate
+	// each group's `series` slice - this allows us to avoid allocating a huge slice if the group only has a few series.
```

Comment:
```suggestion
	// each group's `series` slice (inside `accumulateValue`) - this allows us to avoid allocating a huge slice if the group only has a few series.
```

Reply:
Done: 056f1d9